CCIP Guide - Arbitrary data with acknowledgment of receipt (#1750)
smartcontractkit/documentation: src/content/ccip/tutorials/send-arbitrary-data-receipt-acknowledgment.mdx

## Prerequisites

- This tutorial assumes you have completed the [Send Arbitrary Data](/ccip/tutorials/send-arbitrary-data) tutorial.
- Ensure you have enough testnet funds. This guide requires testnet ETH and LINK on _Ethereum Sepolia_ and testnet MATIC and LINK on _Polygon Mumbai_.

  - Go to [faucets.chain.link](https://faucets.chain.link/) and get testnet ETH and LINK on _Ethereum Sepolia_, and testnet LINK on _Polygon Mumbai_.
  - Go to [faucet.polygon.technology](https://faucet.polygon.technology/) to acquire testnet MATIC.
  - Learn how to [Fund your contract with LINK](/resources/fund-your-contract).

If you need to use different networks, you can find more faucets on the [LINK Token Contracts](/resources/link-token-contracts) page.

## Tutorial

> You will deploy a _message tracker_ contract on the source chain and an _acknowledger_ contract on the destination chain. You will send a _string_ text from the _message tracker_ contract on _Ethereum Sepolia_ to the _acknowledger_ contract on _Polygon Mumbai_. The _acknowledger_ contract receives the text and sends a message back to the _message tracker_ contract to acknowledge the receipt of the text. Upon reception, the _message tracker_ contract reads the message, updates its message tracking status, and emits an event. You pay CCIP fees in LINK.

Reviewer: ```suggestion
In this tutorial, you will deploy a _message tracker_ contract on the source blockchain (Ethereum Sepolia) and an _acknowledger_ on the destination blockchain (Polygon Mumbai). Throughout the tutorial, you will pay for CCIP fees using LINK tokens. Here is a step-by-step breakdown:

1. **Sending and Building a CCIP Message:** You'll initiate and send a message from the _message tracker_ contract on Ethereum Sepolia to the _acknowledger_ contract on Polygon Mumbai. The _message tracker_ contract constructs a CCIP message that encapsulates the string text and establishes a tracking status for this message before sending it off.
1. **Receiving and Acknowledging the Message:** Upon receiving the text, the _acknowledger_ contract on Polygon Mumbai sends back a CCIP message to the _message tracker_ contract as an acknowledgment of receipt.
1. **Updating Tracking Status:** Once the acknowledgment is received, the _message tracker_ contract updates the tracking status of the initial CCIP message and emits an event to signal completion.
```

        style="max-width: 70%;"
      />

   1. Click on **transact** to deploy the contract. MetaMask prompts you to confirm the transaction. Check the transaction details to make sure you are deploying the contract to _Polygon Mumbai_.

   1. After you confirm the transaction, the contract address appears as the second item in the **Deployed Contracts** list. Copy this contract address.

      <ClickToZoom
        src="/images/ccip/tutorials/deployed-acknowledger-mumbai.webp"
        alt="Chainlink CCIP - Deployed Acknowledger Mumbai"
        style="max-width: 70%;"
      />

   1. Open MetaMask and send <CopyText text="1" code/> LINK to the contract address that you copied. Your contract will pay CCIP fees in LINK.

   <ClickToZoom
     src="/images/ccip/tutorials/fund-deployed-acknowledger-sepolia.webp"
     alt="Chainlink CCIP - Fund Deployed Acknowledger Mumbai"
     style="max-width: 70%;"
   />

You now have one _message tracker_ (sender) contract on _Ethereum Sepolia_ and one _acknowledger_ (receiver) contract on _Polygon Mumbai_. You sent `3` LINK to the _message tracker_ contract and `1` LINK to the _acknowledger_ contract to pay the CCIP fees.

> ### Allow chains and senders

Reviewer: Please include this in the steps above like it is done for other tutorials

Both _message tracker_ and _acknowledger_ contracts must allow each other's chains and addresses. This security measure ensures that only the intended contracts can send and receive messages. You can find the chain selector for each network on the [Supported Networks](/ccip/supported-networks) page.

> #### _Message tracker_ contract

Reviewer: Please include this in the steps above like it is done for other tutorials

   | allowlistSender           | The _acknowledger_ contract address                                                                                                                     | Your deployed contract address, true               |
   | allowlistSourceChain      | CCIP Chain identifier of the source blockchain. You can find each network's chain selector on the [ supported networks page ](/ccip/supported-networks) | <CopyText text="12532609583862916517" code/>, true |

   {" "}

   {" "}

   <ClickToZoom
     src="/images/ccip/tutorials/deployed-message-tracker-allowlist.webp"
     alt="Chainlink CCIP - Message Tracker Allowlist Polygon Mumbai and Acknowledger"
     style="max-width: 70%;"
   />

1. Open MetaMask and select the _Ethereum Sepolia_ network.
1. For each function you expanded and filled in the arguments for, click the **transact** button to call the function. MetaMask prompts you to confirm the transaction. Wait for each transaction to succeed before calling the following function.

> #### _Acknowledger_ contract

Reviewer: please include this in the steps above like it is done for other tutorials

       {" "}

       <ClickToZoom
         src="/images/ccip/tutorials/deploy-acknowledger-mumbai.webp"
         alt="Chainlink CCIP - Deploy Acknowledger Mumbai"
         style="max-width: 70%;"
       />

    1. Click **transact** to deploy the contract. MetaMask prompts you to confirm the transaction. Check the transaction details to make sure you are deploying the contract to _Polygon Mumbai_.

    1. After you confirm the transaction, the contract address appears as the second item in the **Deployed Contracts** list. Copy this contract address.

       <ClickToZoom
         src="/images/ccip/tutorials/deployed-acknowledger-mumbai.webp"
         alt="Chainlink CCIP - Deployed Acknowledger Mumbai"
         style="max-width: 70%;"
       />

    1. Open MetaMask and send <CopyText text="1" code/> LINK to the contract address that you copied. Your contract will pay CCIP fees in LINK.

       <ClickToZoom
         src="/images/ccip/tutorials/fund-deployed-acknowledger-sepolia.webp"
         alt="Chainlink CCIP - Fund Deployed Acknowledger Mumbai"
         style="max-width: 70%;"
       />

> 1.  As the _acknowledger_ contract will send messages to and receive messages from the _message tracker_ contract deployed on _Ethereum Sepolia_, you must allow the _Ethereum Sepolia_ chain selector in both _acknowledger_'s allowed destination and source chains. You must also allow the _message tracker_ contract address as a sender. The _message tracker_ contract address is the address you copied when you deployed the _message tracker_ contract on _Ethereum Sepolia_.

Reviewer: ```suggestion
1.  As the _acknowledger_ contract will send messages to and receive messages from the _message tracker_ contract deployed on _Ethereum Sepolia_, you must allow the _Ethereum Sepolia_ chain selector for both destination and source chains. You must also allow the _message tracker_ contract address as a sender.
```

the last part is already explained in the description below

    1. On the **Deploy & Run Transactions** tab in Remix, expand the _acknowledger_ contract in the **Deployed Contracts** section. Expand the `allowlistDestinationChain`, `allowlistSender`, and `allowlistSourceChain` functions and fill in the following arguments:

       | Function                  | Description                                                                                                                                             | Value (_Ethereum Sepolia_)                                                 |
       | ------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------- |
       | allowlistDestinationChain | CCIP Chain identifier of the target blockchain. You can find each network's chain selector on the [ supported networks page ](/ccip/supported-networks) | <CopyText text="16015286601757825753" code/>, <CopyText text="true" code/> |

>        | allowlistSender           | The _message tracker_ contract address                                                                                                                  | Your deployed contract address, <CopyText text="true" code/>               |

Reviewer: ```suggestion
       | allowlistSender           | The address of the message tracker contract deployed on _Ethereum Sepolia_                                                                                                                  | Your deployed contract address, <CopyText text="true" code/>               |
```

         style="max-width: 70%;"
       />

    1. Open MetaMask and select the _Polygon Mumbai_ network.
    1. For each function you expanded and filled in the arguments for, click the **transact** button to call the function. MetaMask prompts you to confirm the transaction. Wait for each transaction to succeed before calling the following function.

1.  Finally, allow your _acknowledger_ contract as a sender in the _message tracker_ contract on _Ethereum Sepolia_:

    1. On the **Deploy & Run Transactions** tab in Remix, expand the _message tracker_ contract in the **Deployed Contracts** section. Expand the `allowlistSender` function and fill in your _acknowledger_ contract address and <CopyText text="true" code/> as allowed.

       {" "}

       <ClickToZoom
         src="/images/ccip/tutorials/deployed-message-tracker-allowlist-sender.webp"
         alt="Chainlink CCIP - Message Tracker Allowlist Acknowledger"
         style="max-width: 70%;"
       />

    1. Open MetaMask and select the _Ethereum Sepolia_ network.
    1. Click **transact** to call the function. MetaMask prompts you to confirm the transaction.

> At this point, you have one _message tracker_ (sender) contract on _Ethereum Sepolia_ and one _acknowledger_ (receiver) contract on _Polygon Mumbai_. You sent `2` LINK to the _message tracker_ contract and `1` LINK to the _acknowledger_ contract to pay the CCIP fees.

Reviewer: @khadni 2 LINK is too much given the destination chain and the gasLimit. You should be able to run the tutorial with 0.5 LINK in Sepolia

- This is the [transaction](https://sepolia.etherscan.io/tx/0x0bcd553f651c3c7c4f3c3abcaa7a2762c6b7ad8fe508da5aca3dd39dd01a33e5) on Ethereum Sepolia: the cost is 0.05 LINK. So with 0.5 LINK , you are already taking some margin.
- This is the [transaction](https://ccip.chain.link/msg/0x390d1cc8061a0e68625d99cacfe5e58f83094e44ddd7e5ae478a31e63299b9df) on Polygon Mumbai.

conclusion, it's fine if we use these fees
- Form sepolia to Mumbai: 0.5 LINK
- From Mumbai to Sepolia: 1 LINK

   />

   Note the returned `status 1`. This value indicates that the _message tracker_ contract has updated your message status to the `Sent` status as defined by the `MessageStatus` `enum` in the _message tracker_ contract.

   ```solidity
   // Enum is used to track the status of messages sent via CCIP.
   // `NotSent` indicates a message has not yet been sent.
   // `Sent` indicates that a message has been sent to the Acknowledger contract but not yet acknowledged.
   // `Confirmed` indicates that the Message Tracker contract has received the acknowledgment from the Acknowledger contract.
   enum MessageStatus {
     NotSent, // 0
     Sent, // 1
     Confirmed // 2
   }
   ```

1. When the transaction is marked with a "Success" status on the [CCIP Explorer](https://ccip.chain.link/), the CCIP transaction and the destination transaction are complete. The _acknowledger_ contract has received the message from the _message tracker_ contract.

   <ClickToZoom
     src="/images/ccip/tutorials/ccip-explorer-sepolia-mumbai-success.webp"
     alt="Chainlink CCIP Explorer - Sepolia to Mumbai Transaction Success"
     style="max-width: 70%;"
   />

#### Acknowledgment message

> The `acknowledgePayLINK` function acts as a custom [`sendMessagePayLINK`](https://docs.chain.link/ccip/tutorials/send-arbitrary-data#sending-data-and-pay-in-link) function nested within the `ccipReceive` function of the _acknowledger_ contract and sends an acknowledgment (a CCIP message) to the _message tracker_ contract upon the initial message receipt. The data transferred in this acknowledgment message is the initial message ID.

Reviewer: It's a bit confusing because you mix up code explanation with the tutorial. Users who haven't read the code explanation yet or the contract code will be confused. I suggest to rework this part:

- Just keep it general and inform that the ccipReceive function processes the message then sends a CCIP message (containing the acknowledgment) to the sender
- When the message is sent, an event is sent (indeed, you can keep the structure of the event because it is necessary)
- Link to the code explanation for more details

  bytes32 data, // The data being sent back, usually containing the message ID of the original message to acknowledge its receipt.
  address feeToken, // The token address used to pay CCIP fees for sending the acknowledgment.
  uint256 fees // The fees paid for sending the acknowledgment message via CCIP.
);
```

1. Copy your _acknowledger_ contract address from Remix. Open the [_Polygon Mumbai_ explorer](https://mumbai.polygonscan.com) and search for your deployed _acknowledger_ contract, then click on the **Events** tab.

   <ClickToZoom
     src="/images/ccip/tutorials/polygon-mumbai-explorer-acknowledger-events.webp"
     alt="Chainlink CCIP - Acknowledger Get Last Received Message Details"
     style="max-width: 70%;"
   />

   The first indexed topic (`topic1`) in the `AcknowledgmentSent` event is the acknowledgment message ID sent to the _message tracker_ contract on _Ethereum Sepolia_. In this example, the message ID is `0x0defcd4d21415378e7136b7300940fb8c0e9e046002aa9b54a0362ecb5cee0ec`.

1. Copy your own message ID from the indexed `topic1` and search for it in the [CCIP explorer](https://ccip.chain.link/).

>    <ClickToZoom

Reviewer: This part will confuse readers. In fact, there is an event MessageReceived, but we cannot see any event AcknowledgmentSent.
At your place, I would have merged the 2 events (basically only keep AcknowledgmentSent and put relevant info of MessageReceived there). Thus, readers will only see one event

     src="/images/ccip/tutorials/ccip-explorer-mumbai-sepolia-success.webp"
     alt="Chainlink CCIP - CCIP Explorer Mumbai to Sepolia Transaction Success"
     style="max-width: 70%;"
   />

   When the transaction is marked with a "Success" status on the CCIP explorer, the CCIP transaction and the destination transaction are complete. The _message tracker_ contract has received the message from the _acknowledger_ contract.

#### Final status check

Upon reception of the acknowledgment message, the _message tracker_ `ccipReceive` function updates the initial message status to `2`, which corresponds to the `Confirmed` status as defined by the `MessageStatus` `enum` and emits a `MessageConfirmed` event.

> 1. On the **Deploy & Run Transactions** tab in Remix, expand your _message tracker_ contract in the **Deployed Contracts** section.

Reviewer: Please ask the users to make sure they are connected to ETH Sepolia

> 1. Paste the message ID you copied from the CCIP explorer as the argument in the **messagesInfo** getter function. Click **messagesInfo** to read the message status. It returns `status 2` and the acknowledger message ID that confirms this status.

Reviewer: Make sure users understand that we are talking here about the first CCIP message ID (so that you don't have users who mistakenly copy the messageID of the message that was sent by the Acknowledger contract)

      | destinationChainSelector | CCIP Chain identifier of the target blockchain. You can find each network's chain selector on the [ supported networks page ](/ccip/supported-networks) | <CopyText text="12532609583862916517" code/>  |
      | receiver                 | The destination smart contract address                                                                                                                  | Your deployed _acknowledger_ contract address |
      | text                     | Any `string`                                                                                                                                            | <CopyText text="Hello World!" code/>          |

      <ClickToZoom
        src="/images/ccip/tutorials/message-tracker-send-message.webp"
        alt="Chainlink CCIP - Message Tracker Send Message to Acknowledger"
        style="max-width: 70%;"
      />

   1. Click **transact** to call the function. MetaMask prompts you to confirm the transaction.

      <Aside type="note">

>         During gas price spikes, your transaction might fail, requiring more than _0.5 LINK_ to proceed. If your

Author: @aelmanaa I'll leave the decision to you, as you may want consistency among the CCIP guides

>         During gas price spikes, your transaction might fail, requiring more than _0.5 LINK_ to proceed. If your

Reviewer 2: @aelmanaa and I had a chat about this. We agree it's better to take this approach for gas prices than to ask users to make more testnet LINK available for each guide. In most situations users won't hit a gas spike.

>         During gas price spikes, your transaction might fail, requiring more than _0.5 LINK_ to proceed. If your

Reviewer: @khadni it's ok to leave it like this for now after discussion with @dwightjl . We will revisit this later

   <ClickToZoom
     src="/images/ccip/tutorials/polygon-mumbai-explorer-acknowledger-events.webp"
     alt="Chainlink CCIP - Polygon Mumbai Acknowledger Contract Events"
     style="max-width: 70%;"
   />

   The first indexed topic (`topic1`) in the `AcknowledgmentSent` event is the acknowledgment message ID sent to the _message tracker_ contract on _Ethereum Sepolia_. In this example, the message ID is `0x0defcd4d21415378e7136b7300940fb8c0e9e046002aa9b54a0362ecb5cee0ec`.

1. Copy your own message ID from the indexed `topic1` and search for it in the [CCIP explorer](https://ccip.chain.link/).

   <ClickToZoom
     src="/images/ccip/tutorials/ccip-explorer-mumbai-sepolia-success.webp"
     alt="Chainlink CCIP - CCIP Explorer Mumbai to Sepolia Transaction Success"
     style="max-width: 70%;"
   />

   When the transaction is marked with a "Success" status on the CCIP explorer, the CCIP transaction and the destination transaction are complete. The _message tracker_ contract has received the message from the _acknowledger_ contract.

#### Final status check

> Upon reception of the acknowledgment message, the _message tracker_ `ccipReceive` function updates the initial message status to `2`, which corresponds to the `Confirmed` status as defined by the `MessageStatus` `enum` and emits a `MessageConfirmed` event.

Reviewer 2: ```suggestion
When the _message tracker_ receives the acknowledgment message, the `ccipReceive` function updates the initial message status to `2`, which corresponds to the `Confirmed` status as defined by the `MessageStatus` `enum`. The function emits a `MessageConfirmed` event.
```
